rir/src/compiler/analysis/verifier.cpp

```diff
@@ -169,32 +169,109 @@ class TheVerifier {
+                    if (input == phi->bb()) {
+                        std::cerr << "Error at instruction '";
+                        i->print(std::cerr);
+                        std::cerr << "': input '";
+                        iv->printRef(std::cerr);
+                        std::cerr << "' has the phi block as input block\n";
+                        ok = false;
+                    }
+
+                    if (slow) {
+                        if (cfg(bb->owner).isImmediatePredecessor(input, bb))
+                            directInputs++;
+
+                        if ((!cfg(bb->owner).isPredecessor(iv->bb(), i->bb()) ||
+                             // A block can be it's own predecessor (loop). But
+                             // then the input must come after the phi!
+                             (iv->bb() == i->bb() &&
+                              i->bb()->indexOf(iv) < i->bb()->indexOf(i)))) {
+                            std::cerr << "Error at instruction '";
+                            i->print(std::cerr);
+                            std::cerr << "': input '";
+                            iv->printRef(std::cerr);
+                            std::cerr
+                                << "' does not come from a predecessor.\n";
+                            ok = false;
+                        }
+                    }
+                }
+            });
+            if (slow) {
+                if (directInputs ==
+                    cfg(bb->owner).immediatePredecessors(bb).size()) {
+                    if (directInputs != phi->nargs()) {
+                        std::cout << "digraph { ";
+                        Visitor::run(f->entry, [&](BB* bb) {
+                            //                          std::cout << "BB" <<
+                            //                          bb->id << "
+                            //                          [shape=\"box\"
+                            //                          label=\"BB" << bb->id <<
+                            //                          "\"]; ";
+                            if (bb->next0)
+                                std::cout << "BB" << bb->id << " -> BB"
+                                          << bb->next0->id << "; ";
+                            if (bb->next1 && !Checkpoint::Cast(bb->last()))
+                                std::cout << "BB" << bb->id << " -> BB"
+                                          << bb->next1->id << "; ";
+                        });
+                        std::cout << "}\n";
+
+                        std::cerr << "Error at instruction '";
+                        i->print(std::cerr);
+                        std::cerr << "'\nI have inputs from all my immediate "
+                                  << " predecessors but "
+                                  << phi->nargs() - directInputs
+                                  << " more arguments than immediate preds.\n";
+                        std::cerr << "\
```

Comment:
I think there's a stupid preprocessor trick you might be able to use to slightly clean this up. Consecutive string literals get concatenated by the preprocessor, so you could write:

```
"This means we created ... \n"
"  .-----.                             \n"
"  | BB1 |         .-----.             \n"
"  |-----|----.----| BB0 |             \n"
"  | a1= |    |    '-----'             \n"
"  '-----'    v                        \n"
```